[MRG + 1] Add check for estimator: parameters not modified by `fit`

kiote opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

Fixes #7763


#### What does this implement/fix? Explain your changes.

Add simple test to estimator_tests, which checks that `__dict__` does not have non-private attributes after `fit`

#### Any other comments?

didn't add any documentation for that, do I need to?
<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

kiote commented on sklearn/utils/estimator_checks.py:
```diff
@@ -454,8 +455,37 @@ def check_dict_unchanged(name, Estimator):
+    if hasattr(estimator, "n_clusters"):
+        estimator.n_clusters = 1
+
+    set_random_state(estimator, 1)
```
this part seems to be very repetitive, so maybe it's possible to refactor / extract method here.. not sure though

amueller replied:
sounds good

kiote commented on sklearn/utils/estimator_checks.py:
```diff
@@ -454,8 +455,37 @@ def check_dict_unchanged(name, Estimator):
+
 def check_fit2d_predict1d(name, Estimator):
-    # check by fitting a 2d array and prediting with a 1d array
+    # check by fitting a 2d array and predicting with a 1d array
```
not relevant to the current commit, just caught my eye. Hope it's innocent enough!

amueller replied:
Sure that's fine :)

kiote commented:
there are some amount of estimators, which is not follow this rule from the beginning, that's the reason of failing tests

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -454,8 +455,37 @@ def check_dict_unchanged(name, Estimator):
+    for val in substracted_dicts:
+        assert_true(val.startswith('_') or val.endswith('_'),
+                    ('Estimator sets invalid attributes during the fit method'
+                     'should either start with an _ or end with a _'))
```
Could you please list the attributes that are set inappropriately here? Then the test logs will be more informative.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -454,8 +455,37 @@ def check_dict_unchanged(name, Estimator):
+    estimator.fit(X, y)
+
+    dict_after_fit = estimator.__dict__
+    # leave only attributes which have been set by fit
```
I think we'd also like to check that attributes set in `__init__` were not modified by `fit`

kiote replied:
okay, that uncovers my total lack of knowledge of how the things work 😕 . I wonder how this sentence and this comment: https://github.com/scikit-learn/scikit-learn/pull/7553#issuecomment-251103120 are working together?

kiote replied:
😿

jnothman replied:
why the tears? Sorry response times are slow atm... I'm watching, but everything is just queued up for the moment!

kiote replied:
haha I understand! Just trying to figure out how to make this run and for me your sentence saying: "attributes set in `__init__` were not modified by `fit`" and  @amueller comment some time ago saying: "We certainly change the `__dict__` during `fit`" conflicting in my mind, so I can't really continue here without resolving this conflict.

So the real question is: should we or shouldn't we modify `__dict__` during `fit` for any given estimator?

Do I understand correctly, that we can add new attributes to `__dict__` during `fit`, but we can't modify the old one (set by `__init__`)?

jnothman replied:
In `fit`, we should only usually be modifying (usually adding, but not necessarily) attributes that start or end with `_`.

kiote commented:
Made error message more informative, also add one more test, checking that `fit` does not change private attributes. So, with this we have two new tests: `fit` does not add new non-private attributes and `fit` does not change any non-private attributes added before.

amueller commented:
the trailing underscore is not about public or private, it's a scikit-learn convention.
*Leading* underscores are indicating private attributes.
In ``__init__`` we should only add attributes that have neigher a leading not a trailing underscore. In ``fit`` we should not change any of these, but we can add (or change if we call fit repeatedly) attributes with trailing underscore. Adding or changing private attributes is also fine.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -454,8 +456,85 @@ def check_dict_unchanged(name, Estimator):
+
+    dict_after_fit = estimator.__dict__
+    # leave only attributes which have been set by fit
+    substracted_dicts_keys = [k for k in dict_after_fit.keys()
```
I think you should check that they are the same before and after.

kiote replied:
yes, this we check this in the previous test [here](https://github.com/kiote/scikit-learn/blob/82091b7bdc867b72ec51d5554e7389412798b385/sklearn/utils/estimator_checks.py#L455)

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -454,8 +456,85 @@ def check_dict_unchanged(name, Estimator):
+    substracted_dicts_keys = [k for k in dict_after_fit.keys()
+                              if k not in dict_before_fit.keys()]
+
+    for val in substracted_dicts_keys:
```
This is the right test :)

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -454,8 +456,85 @@ def check_dict_unchanged(name, Estimator):
+                    if not (attr.startswith('_') or attr.endswith('_'))]
+
+    for attr in public_attrs:
+        assert_equal(dict_before_fit[attr], dict_after_fit[attr],
```
why don't you just do this as part of the test above, instead of removing the entries in the dict? There is a lot of code duplication otherwise as you saw.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -454,8 +456,85 @@ def check_dict_unchanged(name, Estimator):
+
+
+def check_fit_changes_private_attributes_only(name, Estimator):
+    if name in ['GaussianProcess', 'GaussianProcessRegressor',
```
This is pretty bad. Can you remove this line so that we can see the errors in continuous integration? These seems problematic and we might want to fix them.

kiote replied:
removed

amueller reviewed (changes requested):
This looks pretty good already :)

kiote commented:
I uncommented passing of the estimators. So now those ones which do not follow the rule "do not change public attributes" are failing in the CI.

kiote commented:
Should I try to do something with these estimators?

jnothman commented:
These appear to be `KeyError`s, not `AssertionErrors`. Could you please start by making the error message more informative as to what needs to be fixed? Then please list the errors in a comment here so that we can review if they are straightforward to fix.

kiote commented:
here:

| Estimator | Error |
|-----------|------|
| GaussianProcess | Estiamtor adds public attribute during the fit method. Estimators are only allowed to add private attributes either started with _ or ended with _ but X_mean added |
| GaussianProcessRegressor | ... rng added |
| GradientBoostingClassifier | ...  n_features added |
| GradientBoostingRegressor | ... n_features added |
| GraphLassoCV | ... grid_scores added |
| LarsCV | ... fit_path added |
| LassoLarsCV | ... fit_path added |
| LassoLarsIC | ... fit_path added |
| PassiveAggressiveClassifier | ... loss_function added |
| Perceptron | ... loss_function added | 
| SGDClassifier | ... loss_function added |
| TSNE | ... n_iter_final added |

jnothman commented:
Thanks! That is a very useful summary. Will look into it soon.

jnothman commented:
This is great, thanks @kiote, although I can tell that it's only erroring for one attribute even if multiple are set. Would be a good idea to list all added.

In my opinion:

| Estimator | Attribute | Solution |
|-----------|-----------|----------|
| GaussianProcess | X_mean | we could deprecate this and other similar attributes (`X_std`) etc, so that the tests pass, but the class is deprecated in entirety. |
| GaussianProcessRegressor | rng | Deprecate for removal to a local variable, I think |
| GradientBoostingClassifier |  n_features | deprecate and rename to `_n_features`; or abolish it and use `len(self.feature_importances_)` instead at prediction time. |
| GradientBoostingRegressor | n_features | " |
| GraphLassoCV | grid_scores | Deprecate and rename to `grid_scores_`. We may replace this with `cv_results_` some time soon, anyway. |
| LarsCV | fit_path | Move initialisation to `__init__`, I think |
| LassoLarsCV | fit_path | " |
| LassoLarsIC | fit_path | " |
| PassiveAggressiveClassifier | loss_function | Deprecate and rename to `loss_function_` |
| Perceptron | loss_function | " |
| SGDClassifier | loss_function | " |
| TSNE | n_iter_final | deprecate and rename to `n_iter_` |

Individual PRs welcome. If you don't want to make that effort, please post this as one or more new issues so we can seek contributors.

kiote commented:
Thanks Joel!

To make this issue bounded somehow, I can suggest:

1. As you said, add all attributes to error message instead of just one;
2. New input may be needed after that;
3. Make a deprecation message instead of error messages while testing, to be able to merge this PR;
4. Create new issue saying "replace deprecation message with failing tests" or something like that and fix all estimators during that new issue (hopefully I'll be able to do that as well).

What do you think?

jnothman commented:
I'd like to try and fix the issues as far as possible before merging this PR. It shouldn't be too hard to do most of those changes.

kiote commented:
okay than I'm on that

amueller commented:
Maybe skip classes that are deprecated? (you can find out if something is deprecated if ``object.__init__`` has an attribute ``_deprecated_original``.

amueller commented:
If only we had some way to easily rename attributes *cough* futurepast*cough*. Maybe I should finish that at some point?

kiote commented:
After skipping depreciated estimators and printing all attributes we have the same result [as before](https://github.com/scikit-learn/scikit-learn/pull/7846#issuecomment-262709075) but without `GaussianProcess ` and with 

> Estimators are only allowed to add private attributes either started with _ or ended with _ but rng, y_train_mean added 

error for `GaussianProcessRegressor`. I can follow Joel's recommendations or just stop here and wait for  Andreas :suspect:

jnothman commented:
y_train_mean can become _y_train_mean. y_train_ is already stored.

On 2 December 2016 at 00:48, Ekaterina Krivich <notifications@github.com>
wrote:

> After skipping depreciated estimators and printing all attributes we have
> the same result as before
> <https://github.com/scikit-learn/scikit-learn/pull/7846#issuecomment-262709075>
> but without GaussianProcess and with
>
> Estimators are only allowed to add private attributes either started with
> _ or ended with _ but rng, y_train_mean added
>
> error for GaussianProcessRegressor. I can follow Joel's recommendations
> or just stop here and wait for Andreas [image: :suspect:]
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7846#issuecomment-264177031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz68FYzEE8XxbgudDeWKgRu3VseL6pks5rDtAdgaJpZM4KtRdT>
> .
>

amueller commented on sklearn/covariance/graph_lasso_.py:
```diff
@@ -564,6 +565,12 @@ def __init__(self, alphas=4, n_refinements=4, cv=None, tol=1e-4,
         self.store_precision = True
 
+    @property
+    @deprecated("Attribute grid_scores was deprecated in version 0.__ and "
```
0.21

amueller commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -140,8 +141,20 @@ def __init__(self, kernel=None, alpha=1e-10,
+        return self._rng
+
+    @property
+    @deprecated("Attribute y_train_mean was deprecated in version 0.__ and "
```
as above

amueller commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -337,6 +337,12 @@ def __init__(self, loss="hinge", penalty='l2', alpha=0.0001, l1_ratio=0.15,
         self.n_jobs = int(n_jobs)
 
+    @property
+    @deprecated("Attribute loss_function was deprecated in version 0.__ and "
```
again

kiote commented:
Set right versions in the deprecation messages

amueller commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -850,25 +850,26 @@ def _check_params(self):
                 # if is_classification
                 if self.n_classes_ > 1:
-                    max_features = max(1, int(np.sqrt(self.n_features)))
+                    max_features = max(1, int(np.sqrt(self._n_features)))
```
what was the reason we decided to make this private again?

kiote replied:
because before `n_features` was changed by `fit` here: https://github.com/kiote/scikit-learn/blob/0e18dc9141ee97e6e183b5b457f9670e17c8fe8b/sklearn/ensemble/gradient_boosting.py#L981

and now that makes tests to fail with the new checker.

kiote replied:
because before `n_features` was changed by `fit` here: https://github.com/kiote/scikit-learn/blob/0e18dc9141ee97e6e183b5b457f9670e17c8fe8b/sklearn/ensemble/gradient_boosting.py#L981

and now that makes tests to fail with the new checker.

amueller replied:
I meant why do ``_n_features`` instead of ``n_features_``.

jnothman replied:
I recommended it, @amueller, because `n_features_` seems a strange model attribute, especially as it can be derived from `feature_importances_`.

tguillemot replied:
I'm not an expert of that code but it is important to keep `n_features` in memory.
Can we change the signature of `_check_params(self)` -> `_check_params(self, X, y)` ?

`X, y = check_X_y(X, y, accept_sparse=['csr', 'csc', 'coo'], dtype=DTYPE)` can also be done in `_check_params`.

amueller commented on sklearn/linear_model/least_angle.py:
```diff
@@ -1087,6 +1087,7 @@ def __init__(self, fit_intercept=True, verbose=False, max_iter=500,
         self.max_n_alphas = max_n_alphas
         self.n_jobs = n_jobs
         self.eps = eps
+        self.fit_path = True
```
I'm confused. @jnothman is this ok? I guess there's no harm and it doesn't seem to break anything?

amueller replied:
Actually, I don't think there is... it's *interesting* that it's not called anywhere here.. I'm not sure...

jnothman replied:
@kiote, could you please use a call to super's `__init__` here? If nothing breaks, that's my preference.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -483,8 +484,61 @@ def check_dict_unchanged(name, Estimator):
+    return not (attr.startswith('_') or attr.endswith('_'))
+
+
+def check_fit_changes_private_attributes_only(name, Estimator):
```
it can also add / change attributes that have underscores so name and docstring are not really accurate

kiote replied:
hm, can't think about something better than `check_fit_adds_or_changes_underscored_attributes_only`, but that seems strange!

amueller replied:
``check_dont_overwrite_parameters``?

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -483,8 +484,61 @@ def check_dict_unchanged(name, Estimator):
                               'Estimator changes __dict__ during %s' % method)
 
 
+def _is_public(attr):
```
this name is not really appropriate. maybe it should be ``is_init_parameter`` and should return the inverse?

kiote replied:
renamed to `is_public_parameter`, since I use it only once and in "not inversed" version

amueller reviewed (commented):
This looks good :) If you think it's ready for review, please rename it to [MRG] .. instead of [WIP].

amueller commented:
hm why doesn't this catch #7752 ? (with PR at #8015)

kiote commented:
Not really sure I understand the last question, but looks like PR #8015 does exactly the same as I do [here](https://github.com/scikit-learn/scikit-learn/pull/7846/files#diff-f0c531307d44caf8fd88c21014a9f7fdR147). So, shouldn't be too bad.

amueller commented:
@kiote ok, never mind, I overlooked your fix.

amueller commented:
LGTM.
@jnothman what do you think about https://github.com/scikit-learn/scikit-learn/pull/7846#discussion-diff-91332830R1090

jnothman reviewed (commented):
sorry, @amueller, my comment had been pending :\

jnothman commented on sklearn/covariance/graph_lasso_.py:
```diff
@@ -564,6 +565,12 @@ def __init__(self, alphas=4, n_refinements=4, cv=None, tol=1e-4,
         self.store_precision = True
 
+    @property
+    @deprecated("Attribute grid_scores was deprecated in version 0.19 and "
```
Need to update Attribute documentation above to indicate `grid_scores_`

kiote replied:
changed

jnothman commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -850,25 +850,26 @@ def _check_params(self):
                 # if is_classification
                 if self.n_classes_ > 1:
-                    max_features = max(1, int(np.sqrt(self.n_features)))
+                    max_features = max(1, int(np.sqrt(self.n_features_)))
```
If this is to be a public attribute, it should be documented in all applicable classes' docstrings.

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -140,8 +141,20 @@ def __init__(self, kernel=None, alpha=1e-10,
 
+    @property
+    @deprecated("Attribute rng was deprecated in version 0.19 and "
+                "will be removed in 0.21. Use '_rng' instead")
```
again, documentation

jnothman replied:
If it's `_rng`, i.e. private, then don't tell users to use it. Just tell them it's gone.

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -140,8 +141,20 @@ def __init__(self, kernel=None, alpha=1e-10,
+
+    @property
+    @deprecated("Attribute y_train_mean was deprecated in version 0.19 and "
+                "will be removed in 0.21. Use '_y_train_mean' instead")
```
Do not tell the user to use a private attribute. Just say it's deprecated and will be removed.

jnothman commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -337,6 +337,12 @@ def __init__(self, loss="hinge", penalty='l2', alpha=0.0001, l1_ratio=0.15,
         self.n_jobs = int(n_jobs)
 
+    @property
+    @deprecated("Attribute loss_function was deprecated in version 0.19 and "
```
documentation

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -788,6 +789,12 @@ def _fit(self, X, skip_num_points=0):
 
+    @property
+    @deprecated("Attribute n_iter_final was deprecated in version 0.19 and "
+                "will be removed in 0.21. Use 'n_iter_' instead")
```
documentation

jnothman reviewed (changes requested):
With deprecations, we should also make sure that any public, and ideally private including testing, use of these attributes in scikit-learn makes use of the new name.

jnothman commented:
I'm sorry if all these deprecations aren't as exciting as the PR you originally hoped to contribute! :)

jnothman commented:
Thank you

amueller commented:
Oh yeah you definitely need to check if any of the tests raise deprecation warnings. Not sure why I gave this my +1 without that....

kiote committed:
Add check for estimator: parameters not modified by `fit`

  ensure that estimators only add private attributes and attributes with
  trailing _

  in cases when existing estimators don't follow this new rule, we deprecate the
  attributes and make them follow this rule

  see #7763

jnothman commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -953,6 +960,8 @@ def fit(self, X, y, sample_weight=None, monitor=None):
             computing held-out estimates, early stopping, model introspect, and
             snapshoting.
 
+        n_features_ : the number of features.
```
This should appear in an Attributes section of the estimator docstrings, not the Parameters section of the `fit` docstring.

However, I'm still tempted to say it should be a private `_n_features`, not a public `n_features_`

kiote replied:
okay, renamed to `_n_features`

jnothman commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -481,6 +488,8 @@ def partial_fit(self, X, y, classes=None, sample_weight=None):
             Weights applied to individual samples.
             If not provided, uniform weights are assumed.
 
+        loss_function_ : concrete ``LossFunction``
```
Should appear in Attributes section of class docstring.

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -689,6 +690,9 @@ def _fit(self, X, skip_num_points=0):
             `skip_num_points`. This is useful when computing transforms of new
             data where you'd like to keep the old data fixed.
+
+        n_iter_ : int
```
should appear in Attributes section of class docstring

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -689,6 +690,9 @@ def _fit(self, X, skip_num_points=0):
             data where you'd like to keep the old data fixed.
+
+        n_iter_ : int
+            Number of iterations
```
in other places we describe `n_iter_` as "Number of iterations run."

kiote commented:
I've added (just one yet) [test for deprecation message](https://github.com/scikit-learn/scikit-learn/pull/7846/commits/0bc02d129f78faa3d366665bbc4ccdb46f3a4604) as @amueller mentioned [here](https://github.com/scikit-learn/scikit-learn/pull/7846#issuecomment-266489044). Hopefully that's what you mean ;)

jnothman commented on sklearn/covariance/tests/test_graph_lasso.py:
```diff
@@ -131,3 +132,25 @@ def test_graph_lasso_cv(random_state=1):
+        return estimator.grid_scores
+
+    assert_warns_message(DeprecationWarning, depr_message,
+                         get_deprecated_param, graph_lasso)
```
yes, but you should also check that grid_scores and grid_scores_ are equal.

jnothman replied:
This is okay, but FYI, you could replace `get_deprecated_param` with `lambda x: x.grid_scores` and I'd still be happy with the test. Or replace it with `operator.attrgetter('grid_scores')` Or simply replace `get_deprecated_param, graph_lasso` with `lambda: graph_lasso.grid_scores`.

jnothman reviewed (commented):
Yes, that test is moslty good.

kiote commented:
okay, now I can see some tests fails with the reason I can't really explain (but definetly related to what I've done here):

```
======================================================================
FAIL: sklearn.linear_model.tests.test_coordinate_descent.test_lasso_cv
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/testvenv/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/linear_model/tests/test_coordinate_descent.py", line 172, in test_lasso_cv
    clf.mse_path_[5].mean(), significant=2)
  File "/home/travis/testvenv/lib/python3.5/site-packages/numpy/testing/utils.py", line 665, in assert_approx_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal to 2 significant digits:
 ACTUAL: 1.8208200762282036
 DESIRED: 0.9701523969185478
>>  raise AssertionError('\nItems are not equal to 2 significant digits:\n ACTUAL: 1.8208200762282036\n DESIRED: 0.9701523969185478')
    
======================================================================
FAIL: sklearn.linear_model.tests.test_least_angle.test_estimatorclasses_positive_constraint
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/testvenv/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/linear_model/tests/test_least_angle.py", line 490, in test_estimatorclasses_positive_constraint
    assert_true(min(estimator.coef_) >= 0)
AssertionError: False is not true
    'False is not true' = self._formatMessage('False is not true', "%s is not true" % safe_repr(False))
>>  raise self.failureException('False is not true')
    
======================================================================
FAIL: sklearn.utils.tests.test_estimator_checks.test_check_estimator
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/testvenv/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/utils/tests/test_estimator_checks.py", line 151, in test_check_estimator
    check_estimator, ChangesWrongAttribute)
AssertionError: AssertionError not raised by check_estimator
    'AssertionError not raised by check_estimator' = <unittest.case._AssertRaisesContext object at 0x7fb560098fd0>.test_case._formatMessage(<unittest.case._AssertRaisesContext object at 0x7fb560098fd0>.'AssertionError not raised by check_estimator', 'AssertionError not raised by check_estimator')
>>  raise <unittest.case._AssertRaisesContext object at 0x7fb560098fd0>.test_case.failureException('AssertionError not raised by check_estimator')
```

amueller commented on sklearn/covariance/tests/test_graph_lasso.py:
```diff
@@ -131,3 +134,26 @@ def test_graph_lasso_cv(random_state=1):
+        return estimator.grid_scores
+
+    assert_warns_message(DeprecationWarning, depr_message,
+                         get_deprecated_param, graph_lasso)
```
This test is good (and required), though what I actually mean was "see if any of the tests or examples are using ``grid_scores`` and now raising a deprecation warning, and change the use to ``grid_scores_``".

kiote replied:
ah, now I got it! Okay

amueller commented on sklearn/linear_model/least_angle.py:
```diff
@@ -1076,17 +1076,11 @@ def __init__(self, fit_intercept=True, verbose=False, max_iter=500,
         self.max_n_alphas = max_n_alphas
         self.n_jobs = n_jobs
-        self.eps = eps
+        super(LarsCV, self).__init__()
```
you need to pass all the parameters here.

kiote commented:
For some strange reason if I add `super` to `LarsCV` I have this error:

```
======================================================================
FAIL: sklearn.linear_model.tests.test_coordinate_descent.test_lasso_cv
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/testvenv/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/linear_model/tests/test_coordinate_descent.py", line 172, in test_lasso_cv
    clf.mse_path_[5].mean(), significant=2)
  File "/home/travis/testvenv/lib/python3.5/site-packages/numpy/testing/utils.py", line 665, in assert_approx_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal to 2 significant digits:
 ACTUAL: 0.8912453877601253
 DESIRED: 0.9701523969185478
>>  raise AssertionError('\nItems are not equal to 2 significant digits:\n ACTUAL: 0.8912453877601253\n DESIRED: 0.9701523969185478')
```

which disappears with no `super` call

jnothman commented:
`Lars.__init__` sets an instance attribute `method`. `LassoLarsCV` had formerly overwritten this as a *class* attribute, not an instance attribute. I think the correct fix is to never set method as an instance attribute, but always as a class attribute. Wherever it says `self.method = ...` change it to `method = ...` before `__init__` (i.e. in the class) as is used already in `LarsCV` and `LassoLarsCV`.

Will take a look at the rest soon.

kiote commented:
oh! Finally all tests passed! Thanks for that advice, @jnothman

jnothman commented on sklearn/covariance/tests/test_graph_lasso.py:
```diff
@@ -131,3 +134,26 @@ def test_graph_lasso_cv(random_state=1):
+    graph_lasso = GraphLassoCV(alphas=[0.8, 0.5], tol=1e-1, n_jobs=1)
+    graph_lasso.fit(X)
+
+    depr_message = ("Function grid_scores is deprecated; Attribute grid_scores"
```
This isn't really neat, is it? If we want to be picky, we should avoid the `deprecated` decorator, and just use `warnings.warn` directly in the property getter.

kiote replied:
seems like I misunderstand this, are you saying that `grid_scores` property should be deprecated with `warnings.warn`? If so, need to say that I just followed to [this doc](https://github.com/scikit-learn/scikit-learn/blob/master/doc/developers/contributing.rst#deprecation) saying: "If an attribute is to be deprecated, use the decorator deprecated on a property." 😕

jnothman replied:
:\ indeed. I disagree with those docs. "Function x is deprecated" is very confusing to users for a non-function attribute, isn't it?

kiote replied:
Removed the part saying "Function is deprecated" since it's really not a function! I'd move from `@deprecated` decorator to `warnings`, but then I'd prefer the documentation to be fixed as well, so devs won't be thinking "why this stuff doesn't follow the documentation". What do you think?

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -140,8 +141,20 @@ def __init__(self, kernel=None, alpha=1e-10,
         self.random_state = random_state
 
+    @property
+    @deprecated("Attribute rng was deprecated in version 0.19 and "
```
same here, `warn` could be used to avoid boilerplate text about a function

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -151,6 +164,8 @@ def fit(self, X, y):
         y : array-like, shape = (n_samples, [n_output_dims])
             Target values
 
+        _rng : the RandomState singleton used by np.random.
```
If private, it does not deserve documentation. If documented, it should be in the Attributes section of the class docstring, not in the fit docstring.

kiote replied:
removed!

jnothman commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -325,6 +326,12 @@ def __init__(self, loss="hinge", penalty='l2', alpha=0.0001, l1_ratio=0.15,
         self.n_jobs = int(n_jobs)
 
+    @property
+    @deprecated("Attribute loss_function was deprecated in version 0.19 and "
```
same story

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -468,8 +469,61 @@ def check_dict_unchanged(name, Estimator):
+
+    # check that fit doesn't change any public attribute
+    attrs_changed_by_fit = [key for key in public_keys_after_fit
+                            if (dict_before_fit[key] != dict_after_fit[key])]
```
This is going to break wherever an attribute's value is an array (try `if np.ones(2) == np.ones(2): ...`). It's a weaker test, but perhaps we should just use `is not`.

jnothman replied:
Actually, in some ways it's a weaker test, in other ways an unnecessarily strengthened test :\

kiote replied:
I've got the idea about arrays and non-working check in that case, so changed to `is not`. But really didn't get the second idea (about unnecessary strengthened test)

jnothman replied:
By strengthened I mean that if the parameter changes in fit to a value that is equal, but not the same object, as what it was before, then this test will fail.

jnothman reviewed (commented):
This is some very good work, and apart from the nitpicks, LGTM

jnothman commented:
I'd appreciate it if you didn't squash your commits. it makes it hard to see what has changed.

kiote commented:
ouch, sorry! but I don't think I can un-squash it now :(

jnothman commented:
that's alright, but I was hoping to look at the changes quickly, but
instead will need more time.

On 24 December 2016 at 23:25, Ekaterina Krivich <notifications@github.com>
wrote:

> ouch, sorry! but I don't think I can un-squash it now :(
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7846#issuecomment-269082358>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6zsB72tuWEOITyyFn2NB8M4h39u0ks5rLQ8kgaJpZM4KtRdT>
> .
>

jnothman commented on sklearn/covariance/graph_lasso_.py:
```diff
@@ -525,7 +526,7 @@ class GraphLassoCV(GraphLasso):
         All penalization parameters explored.
 
-    `grid_scores`: 2D numpy.ndarray (n_alphas, n_folds)
+    grid_scores_: 2D numpy.ndarray (n_alphas, n_folds)
```
needs space before colon

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -467,8 +468,62 @@ def check_dict_unchanged(name, Estimator):
+
+    # check that fit doesn't add any public attribute
+    assert_true(not attrs_added_by_fit,
+                ('Estiamtor adds public attribute(s) during'
```
*Estimator

jnothman reviewed (commented):
Otherwise LGTM. (Ignoring the fact that I still don't like those deprecation messages using "Function", but have made another issue to deal with that.)

I'd like @amueller or someone else to review again, however.

kiote committed:
Fix typos

kiote commented:
Just went back from Christmas/New Year celebration and hope you've had a lot of fun as well, guys!

Not sure should I wait for "LGTM" from @amueller or it's ready for merge already?

jnothman commented:
@amueller's still settling into the new year too, it seems. Yes, you should wait for him or someone else to give this another check.

kiote commented:
😭

jnothman commented:
> 😭 

Why? We're a bit pressed for personnel at the moment. The merge will come.

tguillemot commented on sklearn/linear_model/least_angle.py:
```diff
@@ -587,14 +587,15 @@ class Lars(LinearModel, RegressorMixin):
     sklearn.decomposition.sparse_encode
 
     """
+    method = 'lar'
```
`self.method` is it used in another place than l678 and l699. 
I don't like defined like that.

tguillemot replied:
I know it is done in other part of the code but is it absolutely necessary to put it here ?

kiote replied:
we were discussing this change before with @jnothman I suppose, that was his idea to move it here, so we'll be able to make new tests. Also he pointed that it's better to define it this way, so I'm a bit confused, honestly.

I didn't understand what l678 and l699, thought it means "line 678", but not sure what should I see on those lines then 😕

tguillemot replied:
Sorry for that, I've missed the discussion indeed. Forget my comment :)

jnothman replied:
It's not wonderful, but I don't think it's this PR's job to fix it. The problem derives from the use of inheritance, but non-use of `super...__init__`. In the current code, `method` is twice set, like this, as a class attribute (in the CV objects), and twice as an instance attribute. It might make more sense to have it as a private attribute, but this approach at least ensures a consistency that is not in master, and that the instance's `__dict__` only has true parameters.

tguillemot replied:
Ok I see. Thanks for the explanation @jnothman.

tguillemot commented on sklearn/linear_model/least_angle.py:
```diff
@@ -828,6 +829,7 @@ class LassoLars(Lars):
     sklearn.decomposition.sparse_encode
 
     """
+    method = 'lasso'
```
I don't like it.

tguillemot replied:
Sorry for that, I've missed the discussion indeed. Forget my comment :)

kiote commented:
Thanks for your review, but honestly I was trying to keep this PR as much specific as possible, so with that changes you've pointed to (like changing the signature of _check_params) it would be much harder :worried:

tguillemot commented:
@kiote We can change that in another PR indeed. Thanks for your answers and your work.

LGTM

jnothman commented:
@tguillemot, did you happen to check that deprecated attributes etc are no longer in use in e.g. examples? I can't remember if I did that check, but it was @alueller's [last stated concern](https://github.com/scikit-learn/scikit-learn/pull/7846#issuecomment-266489044) since he gave his premature +1.

tguillemot commented:
I have missed one  `n_features`  from [partial_dependence.py -- line 261](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/partial_dependence.py#L261)
```
   if gbrt.n_features != X.shape[1]:
        raise ValueError('X.shape[1] does not match gbrt.n_features')
```

Can you replace it @kiote ?

kiote commented:
okay, I also noticed that I promised to rename `_n_features` to `n_features_` to @amueller, but didn't! Sorry, will fix that, too.

kiote committed:
Rename _n_features to n_features_

kiote committed:
Rename n_features to n_features_

jnothman commented:
Thanks a lot, @kiote!

tguillemot commented:
Thx @kiote
